### PR TITLE
[Example] Labor sampling

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Adding the example files for the labor sampling PR #4668, thus depends on that PR. Requires #6033 to run. The run command:
`python train_lightning.py --dataset=ogbn-products --use-uva --cache-size=500000 --batch-dependency=64`

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
Hi Muhammed:

Is the goal of this example
 - showcase of how good the labor sampling is (reproduce a SOTA research)?
 - or teaching our audience about how to use the labor sampling introduced in https://github.com/dmlc/dgl/pull/4668?
Depends on different goal, we need to reshape the code in a different way.

**Comment by mfbalin** [comment]
Hi Steve,

Labor Sampler has a few additional capabilities that I had to remove from #4668 because they were not explained in the original arxiv paper, I am working on a follow-up paper that will illustrate said capabilities. This PR is basically to showcase all of the features of the Labor Sampler.

I will open another PR to add one of the missing Labor sampling capability, then this PR can be considered to be merged. In order to fully illustrate the benefits of LABOR, I need #4341 to be merged as well.

When it comes to which of the two I am aiming for, I believe that Labor sampler is SOTA when it comes to work required to achieve a certain quality of convergence. But I would also like to showcase how to use LABOR with all of its features. So it is kind of both of those things I am aiming for.

I basically used this PR to generate all my plots for the original arxiv paper, so it is needed for people to reproduce my results in the paper, and also in the follow-up paper I am currently working on.

For now, we can mark this PR a draft, in a month or two if the prerequites I need are merged, I will request a review on this PR.

**Comment by frozenbugs** [comment]
Thanks for the explanation, if the PR is draft for now, then let me remove ppl from reviewers, feel free to add us back when you feel ready.

Keep in mind that we have different guidelines for examples with different goal. If you are just checking the code for ppl to reproduce the result in the paper, the code should be in examples/models/name/....

If the example is an advance tutorial to showcase how to use LABOR with all of its features, then we need to design the code organization and pay more attention to the readability to reduce the barriers for other ppl to learn LABOR from your example. We may think about add more clear comments in the code, have a more detailed readme file, e.t.c. Let's make sure work you are working on can actually achieve your goal, so that you don't have your time wasted :)

**Comment by mfbalin** [comment]
@frozenbugs The required PRs on which this depends are all open and ready now. This depends on the kappa feature and the core GPU cache.

File: examples/pytorch/labor/ladies_sampler.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            prob = torch.ones(insg.num_nodes())
+            prob[insg.out_degrees() == 0] = 0
+        return prob, insg
+
+    def select_neighbors(self, prob, num):
+        """
+        seed_nodes : output nodes
+        cand_nodes : candidate nodes.  Must contain all output nodes in @seed_nodes
+        prob : unnormalized probability of each candidate node
+        num : number of neighbors to sample
```
Comment (peizhou001) [inline examples/pytorch/labor/ladies_sampler.py:70]: `num` -> `num_neighbors`?

**Comment by mfbalin** [comment]
What other changes are required to merge this PR? @frozenbugs

This PR includes everything required to reproduce all the results in the Labor paper and also the kappa feature experiments in the cooperative minibatching paper.

**Comment by frozenbugs** [comment]
Oh, thanks for the reminder, I was waiting for the GPUCache PR's building issue get fix.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**